### ISSUE TITLE
Current batch indexing does not interact nicely with varying mixtures

### ISSUE THREAD

**Issue by kothasuhas** [stanford-crfm/levanter#876]
https://github.com/stanford-crfm/levanter/pull/868 implements varying the mixture over training duration. For this, it relies on user-provided step counts. When the dataset is asked for a batch, it compared the requested indices to this threshold.

Unfortunately, I have now realized that the indices supplied to the dataset might not exactly correspond to the specific training run. In a two stage curriculum, this results in data mixture 2 starting slightly before the requested transition point. This can be observed when running two experiments with the exact same mixture for two stages but different transition points: the runs are completely identical except for right before the transition point. Picture attached for two runs where the mixtures are same but the transition point is either 80% through training or 90% through training

<img width="1590" alt="Image" src="https://github.com/user-attachments/assets/cb4c8f3b-c3b8-465d-a851-d38e9287b5ec" />

Two solutions come to mind
1. Change the requested indices to be in the order of training
2. Look into which indices are being requested and manually account for this within the varying mixture dataset

**Comment by dlwh** [comment]
hrm that isn't good but I don't really understand how this happens unless the mixture block sizes aren't lining up?

**Comment by kothasuhas** [comment]
The mixture block sizes are lining up (I assert this inside the code: https://github.com/stanford-crfm/levanter/blob/1d216d1f05db4ce5b3a62da232bb140c3098df7f/src/levanter/data/mixture.py#L67

In some initial debugging, I think I found that during a training run, the indices being requested by the trainer are not perfectly aligned with the actual step count during training? I should make a minimal script to share.

Is it possible that some intermediate steps are being thrown out, or used for eval, when the trainer iterates? If batch i, element j, always requests index i + batch_size + j, then I can't understand why this bug would occur.

**Comment by dlwh** [comment]
so each machine only loads the data it needs, but it should just be batch_size * step + index_within_batch: https://github.com/stanford-crfm/levanter/blob/93a8aa95a06e312439bd457ae326e661fbfc52e2/src/levanter/data/loader.py#L234-L237

**Comment by kothasuhas** [comment]
I confirmed this was a consequence of order of operations. The eval loss was measured after the very first training step with the new data, leading to the slight deviation. This should be a non-issue for practice
